File: src/openjarvis/core/events.py
"""Thread-safe pub/sub event bus for inter-primitive telemetry.

Extends IPW's ``EventRecorder`` into a full publish/subscribe system so that
any primitive can emit events (e.g. ``INFERENCE_END``) and any other primitive can
react without direct coupling.
"""

from __future__ import annotations

import threading
import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Callable, Dict, List, Optional  # noqa: I001

# ---------------------------------------------------------------------------
# Event taxonomy
# ---------------------------------------------------------------------------


class EventType(str, Enum):
    """Supported event categories."""

    INFERENCE_START = "inference_start"
    INFERENCE_END = "inference_end"
    TOOL_CALL_START = "tool_call_start"
    TOOL_CALL_END = "tool_call_end"
    MEMORY_STORE = "memory_store"
    MEMORY_RETRIEVE = "memory_retrieve"
    AGENT_TURN_START = "agent_turn_start"
    AGENT_TURN_END = "agent_turn_end"
    TELEMETRY_RECORD = "telemetry_record"
    TRACE_STEP = "trace_step"
    TRACE_COMPLETE = "trace_complete"
    CHANNEL_MESSAGE_RECEIVED = "channel_message_received"
    CHANNEL_MESSAGE_SENT = "channel_message_sent"
    SECURITY_SCAN = "security_scan"
    SECURITY_ALERT = "security_alert"
    SECURITY_BLOCK = "security_block"
    SCHEDULER_TASK_START = "scheduler_task_start"
    SCHEDULER_TASK_END = "scheduler_task_end"
    BATCH_START = "batch_start"
    BATCH_END = "batch_end"
    # Phase 14 — Agent Hardening & Security
    TOOL_TIMEOUT = "tool_timeout"
    LOOP_GUARD_TRIGGERED = "loop_guard_triggered"
    CAPABILITY_DENIED = "capability_denied"
    TAINT_VIOLATION = "taint_violation"
    # Phase 15 — Workflow, Skills, Sessions
    WORKFLOW_START = "workflow_start"
    WORKFLOW_NODE_START = "workflow_node_start"
    WORKFLOW_NODE_END = "workflow_node_end"
    WORKFLOW_END = "workflow_end"
    SKILL_EXECUTE_START = "skill_execute_start"
    SKILL_EXECUTE_END = "skill_execute_end"
    SESSION_START = "session_start"
    SESSION_END = "session_end"
    # Phase 16 — A2A Protocol
    A2A_TASK_RECEIVED = "a2a_task_received"
    A2A_TASK_COMPLETED = "a2a_task_completed"
    # Phase 22 — Operators
    OPERATOR_TICK_START = "operator_tick_start"
    OPERATOR_TICK_END = "operator_tick_end"
    # Managed agent lifecycle (distinct from OPERATOR_TICK_* for the operator subsystem)
    AGENT_TICK_START = "agent_tick_start"
    AGENT_TICK_END = "agent_tick_end"
    AGENT_TICK_ERROR = "agent_tick_error"
    AGENT_BUDGET_EXCEEDED = "agent_budget_exceeded"
    AGENT_STALL_DETECTED = "agent_stall_detected"
    AGENT_LEARNING_STARTED = "agent_learning_started"
    AGENT_LEARNING_COMPLETED = "agent_learning_completed"
    AGENT_MESSAGE_RECEIVED = "agent_message_received"
    AGENT_CHECKPOINT_SAVED = "agent_checkpoint_saved"
    # Phase 25 — Configuration Optimization
    OPTIMIZE_RUN_START = "optimize_run_start"
    OPTIMIZE_TRIAL_START = "optimize_trial_start"
    OPTIMIZE_TRIAL_END = "optimize_trial_end"
    OPTIMIZE_RUN_END = "optimize_run_end"
    FEEDBACK_RECEIVED = "feedback_received"


@dataclass(slots=True)
class Event:
    """A single event published on the bus."""

    event_type: EventType
    timestamp: float
    data: Dict[str, Any] = field(default_factory=dict)


# Type alias for subscriber callbacks
Subscriber = Callable[[Event], None]


# ---------------------------------------------------------------------------
# EventBus
# ---------------------------------------------------------------------------


class EventBus:
    """Thread-safe publish/subscribe event bus.

    Subscribers are called synchronously in registration order within the
    publishing thread.  An optional *record_history* flag retains all
    published events for later inspection (useful in tests/telemetry).
    """

    def __init__(self, *, record_history: bool = False) -> None:
        self._subscribers: Dict[EventType, List[Subscriber]] = {}
        self._lock = threading.Lock()
        self._record_history = record_history
        self._history: List[Event] = []

    # -- subscribe / unsubscribe --------------------------------------------

    def subscribe(self, event_type: EventType, callback: Subscriber) -> None:
        """Register *callback* to be called whenever *event_type* is published."""
        with self._lock:
            self._subscribers.setdefault(event_type, []).append(callback)

    def unsubscribe(self, event_type: EventType, callback: Subscriber) -> None:
        """Remove *callback* from listeners for *event_type*."""
        with self._lock:
            listeners = self._subscribers.get(event_type, [])
            try:
                listeners.remove(callback)
            except ValueError:
                pass  # Callback already removed — idempotent

    # -- publish ------------------------------------------------------------

    def publish(
        self,
        event_type: EventType,
        data: Optional[Dict[str, Any]] = None,
    ) -> Event:
        """Create and dispatch an event to all subscribers.

        Returns the published ``Event`` instance.
        """
        event = Event(event_type=event_type, timestamp=time.time(), data=data or {})

        with self._lock:
            if self._record_history:
                self._history.append(event)
            listeners = list(self._subscribers.get(event_type, []))

        for callback in listeners:
            callback(event)

        return event

    # -- history ------------------------------------------------------------

    @property
    def history(self) -> List[Event]:
        """Return a copy of all recorded events (empty if recording is off)."""
        with self._lock:
            return list(self._history)

    def clear_history(self) -> None:
        """Discard all recorded events."""
        with self._lock:
            self._history.clear()


# ---------------------------------------------------------------------------
# Module-level singleton
# ---------------------------------------------------------------------------

_bus: Optional[EventBus] = None
_bus_lock = threading.Lock()


def get_event_bus(*, record_history: bool = False) -> EventBus:
    """Return the module-level ``EventBus`` singleton, creating it if needed."""
    global _bus
    with _bus_lock:
        if _bus is None:
            _bus = EventBus(record_history=record_history)
        return _bus


def reset_event_bus() -> None:
    """Replace the singleton with a fresh instance (for tests)."""
    global _bus
    with _bus_lock:
        _bus = None


__all__ = [
    "Event",
    "EventBus",
    "EventType",
    "Subscriber",
    "get_event_bus",
    "reset_event_bus",
]

File: rust/crates/openjarvis-core/src/events.rs
//! Pub/sub event bus for cross-cutting concerns.
//!
//! Rust translation of `src/openjarvis/core/events.py`.

use parking_lot::Mutex;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

// ---------------------------------------------------------------------------
// Event types
// ---------------------------------------------------------------------------

/// All event types published through the event bus.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EventType {
    // Inference
    InferenceStart,
    InferenceEnd,

    // Tools
    ToolCallStart,
    ToolCallEnd,
    ToolTimeout,

    // Memory
    MemoryStore,
    MemoryRetrieve,

    // Agents
    AgentTurnStart,
    AgentTurnEnd,

    // Telemetry
    TelemetryRecord,

    // Traces
    TraceStep,
    TraceComplete,

    // Security
    SecurityScan,
    SecurityAlert,
    SecurityBlock,
    CapabilityDenied,
    TaintViolation,

    // Loop guard
    LoopGuardTriggered,

    // Workflow
    WorkflowStart,
    WorkflowEnd,

    // Skills
    SkillExecuteStart,
    SkillExecuteEnd,

    // Sessions
    SessionStart,
    SessionEnd,

    // Scheduler
    SchedulerTaskStart,
    SchedulerTaskEnd,

    // Operators
    OperatorTickStart,
    OperatorTickEnd,

    // Channels
    ChannelMessageReceived,
    ChannelMessageSent,
}

impl std::fmt::Display for EventType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = serde_json::to_value(self)
            .ok()
            .and_then(|v| v.as_str().map(String::from))
            .unwrap_or_else(|| format!("{self:?}"));
        write!(f, "{s}")
    }
}

impl std::str::FromStr for EventType {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let quoted = format!("\"{s}\"");
        serde_json::from_str(&quoted).map_err(|_| format!("Unknown event type: {s}"))
    }
}

// ---------------------------------------------------------------------------
// Event
// ---------------------------------------------------------------------------

/// A single event published through the event bus.
#[derive(Debug, Clone)]
pub struct Event {
    pub event_type: EventType,
    pub timestamp: f64,
    pub data: HashMap<String, serde_json::Value>,
}

impl Event {
    /// Create a new event with the current timestamp.
    pub fn new(event_type: EventType, data: HashMap<String, serde_json::Value>) -> Self {
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs_f64();
        Self {
            event_type,
            timestamp,
            data,
        }
    }
}

// ---------------------------------------------------------------------------
// Event bus
// ---------------------------------------------------------------------------

/// Callback type for event subscribers.
pub type Subscriber = Arc<dyn Fn(&Event) + Send + Sync>;

/// Thread-safe pub/sub event bus.
///
/// Subscribers register for specific event types and are called synchronously
/// when events of that type are published.
pub struct EventBus {
    subscribers: Mutex<HashMap<EventType, Vec<Subscriber>>>,
    record_history: bool,
    history: Mutex<Vec<Event>>,
}

impl EventBus {
    /// Create a new event bus.
    ///
    /// If `record_history` is `true`, all published events are retained
    /// and can be retrieved via [`Self::history()`].
    pub fn new(record_history: bool) -> Self {
        Self {
            subscribers: Mutex::new(HashMap::new()),
            record_history,
            history: Mutex::new(Vec::new()),
        }
    }

    /// Subscribe to events of a specific type.
    pub fn subscribe(&self, event_type: EventType, callback: Subscriber) {
        let mut subs = self.subscribers.lock();
        subs.entry(event_type).or_default().push(callback);
    }

    /// Publish an event, calling all registered subscribers.
    ///
    /// Returns the published event.
    pub fn publish(
        &self,
        event_type: EventType,
        data: HashMap<String, serde_json::Value>,
    ) -> Event {
        let event = Event::new(event_type, data);

        // Notify subscribers
        let subs = self.subscribers.lock();
        if let Some(callbacks) = subs.get(&event_type) {
            for callback in callbacks {
                callback(&event);
            }
        }

        // Record history if enabled
        if self.record_history {
            self.history.lock().push(event.clone());
        }

        event
    }

    /// Convenience method to publish an event with no data.
    pub fn emit(&self, event_type: EventType) -> Event {
        self.publish(event_type, HashMap::new())
    }

    /// Get the recorded event history.
    pub fn history(&self) -> Vec<Event> {
        self.history.lock().clone()
    }

    /// Clear the recorded event history.
    pub fn clear_history(&self) {
        self.history.lock().clear();
    }

    /// Get the number of subscribers for a given event type.
    pub fn subscriber_count(&self, event_type: EventType) -> usize {
        self.subscribers
            .lock()
            .get(&event_type)
            .map_or(0, |v| v.len())
    }

    /// Remove all subscribers.
    pub fn clear_subscribers(&self) {
        self.subscribers.lock().clear();
    }
}

impl Default for EventBus {
    fn default() -> Self {
        Self::new(false)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::atomic::{AtomicUsize, Ordering};

    #[test]
    fn test_event_type_serde() {
        let et = EventType::InferenceStart;
        let json = serde_json::to_string(&et).unwrap();
        assert_eq!(json, "\"inference_start\"");
        let parsed: EventType = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed, et);
    }

    #[test]
    fn test_event_type_from_str() {
        let et: EventType = "tool_call_start".parse().unwrap();
        assert_eq!(et, EventType::ToolCallStart);
    }

    #[test]
    fn test_publish_subscribe() {
        let bus = EventBus::new(false);
        let counter = Arc::new(AtomicUsize::new(0));
        let counter_clone = counter.clone();

        bus.subscribe(
            EventType::InferenceStart,
            Arc::new(move |_event| {
                counter_clone.fetch_add(1, Ordering::SeqCst);
            }),
        );

        bus.emit(EventType::InferenceStart);
        bus.emit(EventType::InferenceStart);
        bus.emit(EventType::InferenceEnd); // different type, should not fire

        assert_eq!(counter.load(Ordering::SeqCst), 2);
    }

    #[test]
    fn test_event_data() {
        let bus = EventBus::new(false);
        let received_model = Arc::new(Mutex::new(String::new()));
        let rm = received_model.clone();

        bus.subscribe(
            EventType::InferenceStart,
            Arc::new(move |event| {
                if let Some(model) = event.data.get("model").and_then(|v| v.as_str()) {
                    *rm.lock() = model.to_string();
                }
            }),
        );

        let mut data = HashMap::new();
        data.insert("model".into(), serde_json::json!("qwen3:8b"));
        bus.publish(EventType::InferenceStart, data);

        assert_eq!(*received_model.lock(), "qwen3:8b");
    }

    #[test]
    fn test_history_recording() {
        let bus = EventBus::new(true);
        bus.emit(EventType::InferenceStart);
        bus.emit(EventType::InferenceEnd);

        let history = bus.history();
        assert_eq!(history.len(), 2);
        assert_eq!(history[0].event_type, EventType::InferenceStart);
        assert_eq!(history[1].event_type, EventType::InferenceEnd);
    }

    #[test]
    fn test_history_disabled() {
        let bus = EventBus::new(false);
        bus.emit(EventType::InferenceStart);
        assert!(bus.history().is_empty());
    }

    #[test]
    fn test_clear_history() {
        let bus = EventBus::new(true);
        bus.emit(EventType::InferenceStart);
        assert_eq!(bus.history().len(), 1);
        bus.clear_history();
        assert!(bus.history().is_empty());
    }

    #[test]
    fn test_subscriber_count() {
        let bus = EventBus::new(false);
        assert_eq!(bus.subscriber_count(EventType::ToolCallStart), 0);
        bus.subscribe(EventType::ToolCallStart, Arc::new(|_| {}));
        bus.subscribe(EventType::ToolCallStart, Arc::new(|_| {}));
        assert_eq!(bus.subscriber_count(EventType::ToolCallStart), 2);
    }

    #[test]
    fn test_event_timestamp() {
        let event = Event::new(EventType::InferenceStart, HashMap::new());
        assert!(event.timestamp > 0.0);
    }
}
